src/components/OpenSeadragonComponent.js

```diff
@@ -0,0 +1,189 @@
+  const [, forceUpdate] = useReducer(x => x + 1, 0);
+
+  const moveHandler = useDebouncedCallback(useCallback((event) => {
+    /** Shim to provide a mouse-move event coming from the viewer */
+    viewerRef.current?.raiseEvent('mouse-move', event);
+  }, [viewerRef]), 10);
+
+  const onViewportChange = useCallback((event) => {
+    const { viewport } = event.eventSource;
+
+    onUpdateViewport({
+      bounds: viewport.getBounds(),
+      flip: viewport.getFlip(),
+      rotation: viewport.getRotation(),
+      x: Math.round(viewport.centerSpringX.target.value),
+      y: Math.round(viewport.centerSpringY.target.value),
+      zoom: viewport.zoomSpring.target.value,
+    });
+  }, [onUpdateViewport]);
+
+  useEffect(() => {
+    const viewer = viewerRef.current;
+    if (!viewer) return;
+
+    const { viewport } = viewer;
+
+    // @ts-expect-error
+    if (viewerConfig.x && viewerConfig.y
+      && (Math.round(viewerConfig.x) !== Math.round(viewport.centerSpringX.target.value)
+      // @ts-expect-error
+      || Math.round(viewerConfig.y) !== Math.round(viewport.centerSpringY.target.value))) {
+      viewport.panTo(new Openseadragon.Point(viewerConfig.x, viewerConfig.y), false);
+    }
+
+    // @ts-expect-error
+    if (viewerConfig.zoom && viewerConfig.zoom !== viewport.zoomSpring.target.value) {
+      viewport.zoomTo(viewerConfig.zoom, new Openseadragon.Point(viewerConfig.x, viewerConfig.y), false);
+    }
+
+    if (viewerConfig.rotation && viewerConfig.rotation !== viewport.getRotation()) {
+      viewport.setRotation(viewerConfig.rotation);
+    }
+
+    if (viewerConfig.flip !== undefined && (viewerConfig.flip || false) !== viewport.getFlip()) {
+      viewport.setFlip(viewerConfig.flip);
+    }
+
+    if (viewerConfig.bounds && !viewerConfig.x && !viewerConfig.y && !viewerConfig.zoom) {
+      const rect = new Openseadragon.Rect(...viewerConfig.bounds);
+      if (rect.equals(viewport.getBounds())) {
+        viewport.fitBounds(rect, false);
+      }
+    }
+  }, [viewerConfig, viewerRef]);
+
+  // initialize OSD stuff when this component is mounted
+  useEffect(() => {
+    const viewer = Openseadragon({
+      id,
+      ...osdConfig,
+    });
+
+    viewer.addHandler('canvas-drag', () => {
+      setGrabbing(true);
+    });
+
+    viewer.addHandler('canvas-drag-end', () => {
+      setGrabbing(false);
+    });
+
+    viewer.addHandler('canvas-double-click', ({ position, shift }) => {
+      if (!osdConfig.zoomPerDoubleClick) return;
+
+      const currentZoom = viewer.viewport.getZoom();
+      const zoomRatio = (shift ? 1.0 / osdConfig.zoomPerDoubleClick : osdConfig.zoomPerDoubleClick);
+      viewer.viewport.zoomTo(currentZoom * zoomRatio, viewer.viewport.pointFromPixel(position), false);
+    });
+
+    viewer.addHandler('animation-finish', onViewportChange);
+    // @ts-expect-error
+    viewer.innerTracker.moveHandler = moveHandler;
+
+    viewerRef.current = viewer;
+    setViewer(viewer);
```

Comment:
I don't understand if setViewer is actually doing anything, since you are using a Context now.